okhttp/src/main/java/okhttp3/HttpUrl.java

```diff
@@ -13,1331 +13,1173 @@
-  final String scheme;
-
-  /** Decoded username. */
-  private final String username;
-
-  /** Decoded password. */
-  private final String password;
-
-  /** Canonical hostname. */
-  final String host;
-
-  /** Either 80, 443 or a user-specified port. In range [1..65535]. */
-  final int port;
+class HttpUrl internal constructor(builder: Builder) {
+
+  /** Either "http" or "https".  */
+  internal val scheme: String = builder.scheme ?: throw IllegalStateException("scheme == null")
+
+  /** Decoded username.  */
+  private val username: String = percentDecode(builder.encodedUsername, false)
+
+  /** Decoded password.  */
+  private val password: String = percentDecode(builder.encodedPassword, false)
+
+  /** Canonical hostname.  */
+  internal val host: String = builder.host ?: throw IllegalStateException("host == null")
+
+  /** Either 80, 443 or a user-specified port. In range [1..65535].  */
+  internal val port: Int = builder.effectivePort()
 
   /**
    * A list of canonical path segments. This list always contains at least one element, which may be
    * the empty string. Each segment is formatted with a leading '/', so if path segments were ["a",
    * "b", ""], then the encoded path would be "/a/b/".
    */
-  private final List<String> pathSegments;
+  private val pathSegments: List<String> =
+      percentDecode(builder.encodedPathSegments, false) as List<String>
 
   /**
    * Alternating, decoded query names and values, or null for no query. Names may be empty or
    * non-empty, but never null. Values are null if the name has no corresponding '=' separator, or
    * empty, or non-empty.
    */
-  private final @Nullable List<String> queryNamesAndValues;
-
-  /** Decoded fragment. */
-  private final @Nullable String fragment;
-
-  /** Canonical URL. */
-  private final String url;
-
-  HttpUrl(Builder builder) {
-    this.scheme = builder.scheme;
-    this.username = percentDecode(builder.encodedUsername, false);
-    this.password = percentDecode(builder.encodedPassword, false);
-    this.host = builder.host;
-    this.port = builder.effectivePort();
-    this.pathSegments = percentDecode(builder.encodedPathSegments, false);
-    this.queryNamesAndValues = builder.encodedQueryNamesAndValues != null
-        ? percentDecode(builder.encodedQueryNamesAndValues, true)
-        : null;
-    this.fragment = builder.encodedFragment != null
-        ? percentDecode(builder.encodedFragment, false)
-        : null;
-    this.url = builder.toString();
-  }
+  private val queryNamesAndValues: List<String?>? =
+      builder.encodedQueryNamesAndValues?.let { percentDecode(it, true) }
+
+  /** Decoded fragment.  */
+  private val fragment: String? = builder.encodedFragment?.let { percentDecode(it, false) }
+
+  /** Canonical URL.  */
+  private val url: String = builder.toString()
 
-  /** Returns this URL as a {@link URL java.net.URL}. */
-  public URL url() {
+  val isHttps: Boolean get() = scheme == "https"
```

Comment:
`scheme` is a `val`, so this can just be a precomputed property without the `get()`

Reply:
Fixed.